File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/verify/VerifyEmailServiceTest.java
package uk.gov.justice.digital.hmpps.oauth2server.verify;

import com.microsoft.applicationinsights.TelemetryClient;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.jdbc.core.JdbcTemplate;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken.TokenType;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserTokenRepository;
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.Staff;
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.StaffUserAccount;
import uk.gov.justice.digital.hmpps.oauth2server.security.NomisUserService;
import uk.gov.justice.digital.hmpps.oauth2server.verify.VerifyEmailService.VerifyEmailException;
import uk.gov.service.notify.NotificationClientApi;
import uk.gov.service.notify.NotificationClientException;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class VerifyEmailServiceTest {
    @Mock
    private UserRepository userRepository;
    @Mock
    private UserTokenRepository userTokenRepository;
    @Mock
    private NomisUserService nomisUserService;
    @Mock
    private JdbcTemplate jdbcTemplate;
    @Mock
    private TelemetryClient telemetryClient;
    @Mock
    private NotificationClientApi notificationClient;
    @Mock
    private ReferenceCodesService referenceCodesService;
    @Captor
    private ArgumentCaptor<Map<String, String>> mapCaptor;

    private VerifyEmailService verifyEmailService;

    @Before
    public void setUp() {
        verifyEmailService = new VerifyEmailService(userRepository, userTokenRepository, nomisUserService, jdbcTemplate, telemetryClient, notificationClient, referenceCodesService, "templateId");
    }

    @Test
    public void getEmail() {
        final var user = User.builder().username("bob").email("joe@bob.com").build();
        when(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user));
        final var userOptional = verifyEmailService.getEmail("user");
        assertThat(userOptional).get().isEqualTo(user);
    }

    @Test
    public void getEmail_NoEmailSet() {
        final var user = User.of("bob");
        when(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user));
        final var userOptionalOptional = verifyEmailService.getEmail("user");
        assertThat(userOptionalOptional).isEmpty();
    }

    @Test
    public void isNotVerified_userMissing() {
        when(userRepository.findByUsername(anyString())).thenReturn(Optional.empty());
        assertThat(verifyEmailService.isNotVerified("user")).isTrue();
        verify(userRepository).findByUsername("user");
    }

    @Test
    public void isNotVerified_userFoundNotVerified() {
        final var user = User.of("bob");
        when(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user));
        assertThat(verifyEmailService.isNotVerified("user")).isTrue();
    }

    @Test
    public void isNotVerified_userFoundVerified() {
        final var user = User.builder().username("bob").email("joe@bob.com").build();
        user.setVerified(true);
        when(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user));
        assertThat(verifyEmailService.isNotVerified("user")).isFalse();
    }

    @Test
    public void requestVerification_firstNameMissing() throws NotificationClientException, VerifyEmailException {
        when(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(Boolean.TRUE);
        final var verification = verifyEmailService.requestVerification("user", "email@john.com", "url");
        verify(notificationClient).sendEmail(eq("templateId"), eq("email@john.com"), mapCaptor.capture(), eq(null));
        final var params = mapCaptor.getValue();
        assertThat(params).containsEntry("firstName", "user").containsEntry("verifyLink", verification);
    }

    @Test
    public void requestVerification_firstNamePresent() throws NotificationClientException, VerifyEmailException {
        when(nomisUserService.getNomisUserByUsername(anyString())).thenReturn(Optional.of(getStaffUserAccountForBob()));
        when(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(Boolean.TRUE);
        final var verification = verifyEmailService.requestVerification("user", "eMail@john.COM", "url");
        verify(notificationClient).sendEmail(eq("templateId"), eq("email@john.com"), mapCaptor.capture(), eq(null));
        final var params = mapCaptor.getValue();
        assertThat(params).containsEntry("firstName", "Bob").containsEntry("verifyLink", verification);
    }

    @Test
    public void requestVerification_existingToken() throws NotificationClientException, VerifyEmailException {
        final var user = User.of("someuser");
        when(userRepository.findByUsername("user")).thenReturn(Optional.of(user));
        final var existingUserToken = user.createToken(TokenType.VERIFIED);
        when(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(Boolean.TRUE);
        verifyEmailService.requestVerification("user", "email@john.com", "url");
        assertThat(user.getTokens()).hasSize(1).extracting(UserToken::getToken).doesNotContain(existingUserToken.getToken());
    }

    @Test
    public void requestVerification_verifyToken() throws NotificationClientException, VerifyEmailException {
        final var user = User.of("someuser");
        when(userRepository.findByUsername("user")).thenReturn(Optional.of(user));
        when(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(Boolean.TRUE);
        final var verification = verifyEmailService.requestVerification("user", "email@john.com", "url");
        final var value = user.getTokens().stream().findFirst().orElseThrow();
        assertThat(verification).isEqualTo("url" + value.getToken());
    }

    @Test
    public void requestVerification_saveEmail() throws NotificationClientException, VerifyEmailException {
        final var user = User.of("someuser");
        when(userRepository.findByUsername("user")).thenReturn(Optional.of(user));
        when(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(Boolean.TRUE);
        verifyEmailService.requestVerification("user", "eMail@john.COM", "url");
        verify(userRepository).save(user);
        assertThat(user.getEmail()).isEqualTo("email@john.com");
        assertThat(user.isVerified()).isFalse();
    }

    @Test
    public void requestVerification_sendFailure() throws NotificationClientException {
        when(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(Boolean.TRUE);
        when(notificationClient.sendEmail(anyString(), anyString(), anyMap(), isNull())).thenThrow(new NotificationClientException("message"));

        assertThatThrownBy(() -> verifyEmailService.requestVerification("user", "email@john.com", "url")).hasMessage("message");
    }

    @Test
    public void requestVerification_formatEmailInput() throws NotificationClientException, VerifyEmailException {
        when(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(Boolean.TRUE);
        verifyEmailService.requestVerification("user", "some.u’ser@SOMEwhere.COM", "url");
        verify(notificationClient).sendEmail(eq("templateId"), eq("some.u'ser@somewhere.com"), anyMap(), eq(null));
    }

    @Test
    public void verifyEmail_NoAtSign() {
        verifyEmailFailure("a", "format");
    }

    @Test
    public void verifyEmail_MultipleAtSigns() {
        verifyEmailFailure("a@b.fred@joe.com", "at");
    }

    @Test
    public void verifyEmail_NoExtension() {
        verifyEmailFailure("a@bee", "format");
    }

    @Test
    public void verifyEmail_FirstLastStopFirst() {
        verifyEmailFailure(".a@bee.com", "firstlast");
    }

    @Test
    public void verifyEmail_FirstLastStopLast() {
        verifyEmailFailure("a@bee.com.", "firstlast");
    }

    @Test
    public void verifyEmail_FirstLastAtFirst() {
        verifyEmailFailure("@a@bee.com", "firstlast");
    }

    @Test
    public void verifyEmail_FirstLastAtLast() {
        verifyEmailFailure("a@bee.com@", "firstlast");
    }

    @Test
    public void verifyEmail_TogetherAtBefore() {
        verifyEmailFailure("a@.com", "together");
    }

    @Test
    public void verifyEmail_TogetherAtAfter() {
        verifyEmailFailure("a.@joe.com", "together");
    }

    @Test
    public void verifyEmail_White() {
        verifyEmailFailure("a@be\te.com", "white");
    }

    @Test
    public void verifyEmail_InvalidDomain() {
        verifyEmailFailure("a@b.com", "domain");
        verify(referenceCodesService).isValidEmailDomain("b.com");
    }

    @Test
    public void verifyEmail_InvalidCharacters() {
        verifyEmailFailure("a@b.&com", "characters");
    }

    private void verifyEmailFailure(final String email, final String reason) {
        assertThatThrownBy(() -> verifyEmailService.validateEmailAddress(email)).
                isInstanceOf(VerifyEmailException.class).extracting("reason").isEqualTo(reason);
    }

    private StaffUserAccount getStaffUserAccountForBob() {
        final var staffUserAccount = new StaffUserAccount();
        final var staff = new Staff();
        staff.setFirstName("bOb");
        staffUserAccount.setStaff(staff);
        return staffUserAccount;
    }

    @Test
    public void confirmEmail_happyPath() {
        final var user = User.of("bob");
        final var userToken = user.createToken(TokenType.VERIFIED);
        when(userTokenRepository.findById(anyString())).thenReturn(Optional.of(userToken));
        final var result = verifyEmailService.confirmEmail("token");

        assertThat(result).isEmpty();
        verify(userRepository).save(user);
        assertThat(user.isVerified()).isTrue();
    }

    @Test
    public void confirmEmail_invalid() {
        final var result = verifyEmailService.confirmEmail("bob");
        assertThat(result).get().isEqualTo("invalid");
    }

    @Test
    public void confirmEmail_expired() {
        final var user = User.of("bob");
        final var userToken = user.createToken(TokenType.VERIFIED);
        userToken.setTokenExpiry(LocalDateTime.now().minusSeconds(1));
        when(userTokenRepository.findById(anyString())).thenReturn(Optional.of(userToken));
        final var result = verifyEmailService.confirmEmail("token");
        assertThat(result).get().isEqualTo("expired");
    }
}

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/verify/VerifyEmailServiceTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.verify

import com.microsoft.applicationinsights.TelemetryClient
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.isNull
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyMap
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.verify
import org.springframework.jdbc.core.JdbcTemplate
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Contact
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.ContactType.SECONDARY_EMAIL
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.Person
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken.TokenType
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserTokenRepository
import uk.gov.justice.digital.hmpps.oauth2server.verify.VerifyEmailService.VerifyEmailException
import uk.gov.service.notify.NotificationClientApi
import uk.gov.service.notify.NotificationClientException
import java.time.LocalDateTime
import java.util.*
import javax.persistence.EntityNotFoundException

class VerifyEmailServiceTest {
  private val userRepository: UserRepository = mock()
  private val userTokenRepository: UserTokenRepository = mock()
  private val jdbcTemplate: JdbcTemplate = mock()
  private val telemetryClient: TelemetryClient = mock()
  private val notificationClient: NotificationClientApi = mock()
  private val referenceCodesService: ReferenceCodesService = mock()
  private val verifyEmailService = VerifyEmailService(userRepository, userTokenRepository, jdbcTemplate, telemetryClient, notificationClient, referenceCodesService, "templateId")

  @Test
  fun email() {
    val user = User.builder().username("bob").email("joe@bob.com").build()
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    val userOptional = verifyEmailService.getEmail("user")
    assertThat(userOptional).get().isEqualTo(user)
  }

  @Test
  fun email_NoEmailSet() {
    val user = User.of("bob")
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    val userOptionalOptional = verifyEmailService.getEmail("user")
    assertThat(userOptionalOptional).isEmpty
  }

  @Test
  fun isNotVerified_userMissing() {
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.empty())
    assertThat(verifyEmailService.isNotVerified("user")).isTrue()
    verify(userRepository).findByUsername("user")
  }

  @Test
  fun isNotVerified_userFoundNotVerified() {
    val user = User.of("bob")
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    assertThat(verifyEmailService.isNotVerified("user")).isTrue()
  }

  @Test
  fun isNotVerified_userFoundVerified() {
    val user = User.builder().username("bob").email("joe@bob.com").build()
    user.isVerified = true
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    assertThat(verifyEmailService.isNotVerified("user")).isFalse()
  }

  @Test
  fun requestVerification_existingToken() {
    val user = User.of("someuser")
    val existingUserToken = user.createToken(TokenType.VERIFIED)
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    verifyEmailService.requestVerification("user", "email@john.com", "firstname", "full name", "url", User.EmailType.PRIMARY)
    assertThat(user.tokens).hasSize(1).extracting<String> { it.token }.doesNotContain(existingUserToken.token)
  }

  @Test
  fun requestVerificationSecondaryEmail_existingToken() {
    val user = User.of("someuser")
    val existingUserToken = user.createToken(TokenType.SECONDARY)
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    verifyEmailService.requestVerification("user", "email@john.com", "firstname", "full name", "url", User.EmailType.SECONDARY)
    assertThat(user.tokens).hasSize(1).extracting<String> { it.token }.doesNotContain(existingUserToken.token)
  }

  @Test
  fun requestVerification_verifyToken() {
    val user = User.of("someuser")
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    val verification = verifyEmailService.requestVerification("user", "email@john.com", "full name", "firstname", "url", User.EmailType.PRIMARY)
    val value = user.tokens.stream().findFirst().orElseThrow()
    assertThat(verification).isEqualTo("url" + value.token)
  }

  @Test
  fun requestVerification_verifyToken_second() {
    val user = User.of("someuser")
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    val verification = verifyEmailService.requestVerification("user", "email@john.com", "firstname", "full name", "url", User.EmailType.SECONDARY)
    val value = user.tokens.stream().findFirst().orElseThrow()
    assertThat(verification).isEqualTo("url" + value.token)
  }

  @Test
  fun requestVerification_saveEmail() {
    val user = User.of("someuser")
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    verifyEmailService.requestVerification("user", "eMail@john.COM", "firstname", "full name", "url", User.EmailType.PRIMARY)
    verify(userRepository).save(user)
    assertThat(user.email).isEqualTo("email@john.com")
    assertThat(user.isVerified).isFalse()
  }

  @Test
  fun requestVerification_sendFailure() {
    val user = User.of("someuser")
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    whenever(notificationClient.sendEmail(anyString(), anyString(), anyMap<String, Any?>(), isNull())).thenThrow(NotificationClientException("message"))
    assertThatThrownBy { verifyEmailService.requestVerification("user", "email@john.com", "firstname", "full name", "url", User.EmailType.PRIMARY) }.hasMessage("message")
  }

  @Test
  fun requestVerification_formatEmailInput() {
    val user = Optional.of(User.of("someuser"))
    whenever(userRepository.findByUsername(anyString())).thenReturn(user)
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    verifyEmailService.requestVerification("user", "some.u’ser@SOMEwhere.COM", "firstname", "full name", "url", User.EmailType.PRIMARY)
    verify(notificationClient).sendEmail(eq("templateId"), eq("some.u'ser@somewhere.com"), anyMap<String, Any?>(), isNull())
  }

  @Test
  fun requestVerification_gsiEmail() {
    val user = Optional.of(User.of("someuser"))
    whenever(userRepository.findByUsername(anyString())).thenReturn(user)
    whenever(referenceCodesService.isValidEmailDomain(anyString())).thenReturn(true)
    verifyPrimaryEmailFailure("some.u'ser@SOMEwhe.gsi.gov.uk", "gsi")
  }

  @Test
  fun verifyEmail_NoAtSign() {
    verifyPrimaryEmailFailure("a", "format")
    verifySecondaryEmailFailure("a", "format")
  }

  @Test
  fun verifyEmail_MultipleAtSigns() {
    verifyPrimaryEmailFailure("a@b.fred@joe.com", "at")
    verifySecondaryEmailFailure("a@b.fred@joe.com", "at")
  }

  @Test
  fun verifyEmail_NoExtension() {
    verifyPrimaryEmailFailure("a@bee", "format")
    verifySecondaryEmailFailure("a@bee", "format")
  }

  @Test
  fun verifyEmail_FirstLastStopFirst() {
    verifyPrimaryEmailFailure(".a@bee.com", "firstlast")
    verifySecondaryEmailFailure(".a@bee.com", "firstlast")
  }

  @Test
  fun verifyEmail_FirstLastStopLast() {
    verifyPrimaryEmailFailure("a@bee.com.", "firstlast")
    verifySecondaryEmailFailure("a@bee.com.", "firstlast")
  }

  @Test
  fun verifyEmail_FirstLastAtFirst() {
    verifyPrimaryEmailFailure("@a@bee.com", "firstlast")
    verifySecondaryEmailFailure("@a@bee.com", "firstlast")
  }

  @Test
  fun verifyEmail_FirstLastAtLast() {
    verifyPrimaryEmailFailure("a@bee.com@", "firstlast")
    verifySecondaryEmailFailure("a@bee.com@", "firstlast")
  }

  @Test
  fun verifyEmail_TogetherAtBefore() {
    verifyPrimaryEmailFailure("a@.com", "together")
    verifySecondaryEmailFailure("a@.com", "together")
  }

  @Test
  fun verifyEmail_TogetherAtAfter() {
    verifyPrimaryEmailFailure("a.@joe.com", "together")
    verifySecondaryEmailFailure("a.@joe.com", "together")
  }

  @Test
  fun verifyEmail_White() {
    verifyPrimaryEmailFailure("a@be\te.com", "white")
    verifySecondaryEmailFailure("a@be\te.com", "white")
  }

  @Test
  fun verifyEmail_InvalidCharacters() {
    verifyPrimaryEmailFailure("a@b.&com", "characters")
    verifySecondaryEmailFailure("a@b.&com", "characters")
  }

  @Test
  fun verifyEmail_InvalidDomain() {
    verifyPrimaryEmailFailure("a@b.com", "domain")
    verify(referenceCodesService).isValidEmailDomain("b.com")
  }

  private fun verifyPrimaryEmailFailure(email: String, reason: String) {
    assertThatThrownBy { verifyEmailService.validateEmailAddress(email, User.EmailType.PRIMARY) }.isInstanceOf(VerifyEmailException::class.java).extracting("reason").isEqualTo(reason)
  }

  private fun verifySecondaryEmailFailure(email: String, reason: String) {
    assertThatThrownBy { verifyEmailService.validateEmailAddress(email, User.EmailType.SECONDARY) }.isInstanceOf(VerifyEmailException::class.java).extracting("reason").isEqualTo(reason)
  }

  @Test
  fun `resendVerificationCodeSecondaryEmail send code`() {
    val user = User.builder().person(Person("bob", "last")).contacts(setOf(Contact(SECONDARY_EMAIL, "someemail", false))).build()
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))

    val verifyLink = verifyEmailService.resendVerificationCodeSecondaryEmail("bob", "http://some.url?token=")
    val token = user.tokens.first().token
    assertThat(verifyLink).get().isEqualTo("http://some.url?token=$token")

    verify(notificationClient).sendEmail(eq("templateId"), eq("someemail"), anyMap<String, Any?>(), isNull())

  }

  @Test
  fun `resendVerificationCodeSecondaryEmail no second email`() {
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(User()))
    assertThatThrownBy { verifyEmailService.resendVerificationCodeSecondaryEmail("bob", "http://some.url") }
        .isInstanceOf(VerifyEmailException::class.java).extracting("reason").isEqualTo("nosecondaryemail")
  }

  @Test
  fun `resendVerificationCodeSecondaryEmail already verified`() {
    val user = User.builder().contacts(setOf(Contact(SECONDARY_EMAIL, "someemail", true))).build()
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    assertThat(verifyEmailService.resendVerificationCodeSecondaryEmail("bob", "http://some.url")).isEqualTo(Optional.empty<String>())
  }

  @Test
  fun `secondaryEmailVerified returns true`() {
    val user = User.builder().contacts(setOf(Contact(SECONDARY_EMAIL, "someemail", true))).build()
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    assertThat(verifyEmailService.secondaryEmailVerified("bob")).isTrue()
  }

  @Test
  fun `secondaryEmailVerified returns false`() {
    val user = User.builder().contacts(setOf(Contact(SECONDARY_EMAIL, "someemail", false))).build()
    whenever(userRepository.findByUsername(anyString())).thenReturn(Optional.of(user))
    assertThat(verifyEmailService.secondaryEmailVerified("bob")).isFalse()
  }

  @Test
  fun `secondaryEmailVerified throws EntityNotFoundException`() {
    assertThatThrownBy { verifyEmailService.secondaryEmailVerified("bob") }
        .isInstanceOf(EntityNotFoundException::class.java).hasMessageContaining("User not found with username bob")
  }

  @Test
  fun confirmEmail_happyPath() {
    val user = User.of("bob")
    val userToken = user.createToken(TokenType.VERIFIED)
    whenever(userTokenRepository.findById(anyString())).thenReturn(Optional.of(userToken))
    val result = verifyEmailService.confirmEmail("token")
    assertThat(result).isEmpty
    verify(userRepository).save(user)
    assertThat(user.isVerified).isTrue()
  }

  @Test
  fun confirmSecondaryEmail_happyPath() {
    val user = User.of("bob")
    val userToken = user.createToken(TokenType.SECONDARY)
    whenever(userTokenRepository.findById(anyString())).thenReturn(Optional.of(userToken))
    val result = verifyEmailService.confirmEmail("token")
    assertThat(result).isEmpty
    verify(userRepository).save(user)
    assertThat(user.isVerified).isTrue()
  }

  @Test
  fun confirmEmail_invalid() {
    val result = verifyEmailService.confirmEmail("bob")
    assertThat(result).get().isEqualTo("invalid")
  }

  @Test
  fun confirmEmail_expired() {
    val user = User.of("bob")
    val userToken = user.createToken(TokenType.VERIFIED)
    userToken.tokenExpiry = LocalDateTime.now().minusSeconds(1)
    whenever(userTokenRepository.findById(anyString())).thenReturn(Optional.of(userToken))
    val result = verifyEmailService.confirmEmail("token")
    assertThat(result).get().isEqualTo("expired")
  }

  @Test
  fun confirmSecondaryEmail_expired() {
    val user = User.of("bob")
    val userToken = user.createToken(TokenType.SECONDARY)
    userToken.tokenExpiry = LocalDateTime.now().minusSeconds(1)
    whenever(userTokenRepository.findById(anyString())).thenReturn(Optional.of(userToken))
    val result = verifyEmailService.confirmEmail("token")
    assertThat(result).get().isEqualTo("expired")
  }
}
